FindMKL.cmake

```diff
@@ -1,230 +1,308 @@
+          PATH_SUFFIXES ${_mkl_libpath_suffix})
+    mark_as_advanced(${_varname}_DYN)
+    find_library(${_varname}_ST
+          NAMES ${_mkl_libname_prefix}${_libname}${_mkl_static_lib}
+          HINTS ${_mkl_search_paths}
+          PATH_SUFFIXES ${_mkl_libpath_suffix})
+    mark_as_advanced(${_varname}_ST)
+endfunction()
 
-# Find MKL header
+# Find MKL headers
 #
 find_path(MKL_INCLUDE_DIR mkl.h
     HINTS ${MKL_ROOT}/include
-    )
+          ${MKL_ROOT}/mkl/include)
 mark_as_advanced(MKL_INCLUDE_DIR)
 
-# Find MKL core libraries
+# Group flags for static libraries on Linux GNU
+#
+if(NOT APPLE AND (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR
+                  CMAKE_CXX_COMPILER_ID STREQUAL "GNU" OR
+                  CMAKE_Fortran_COMPILER_ID STREQUAL "GNU"))
+    set(_mkl_linker_pre_flags_ST "-Wl,--start-group")
+    set(_mkl_linker_post_flags_ST "-Wl,--end-group")
+endif()
+
+# Core MKL
 #
 __mkl_find_library(MKL_CORE_LIB mkl_core)
-__mkl_find_library(MKL_INTERFACE_LIB mkl_intel_${_mkl_lp})
-
-if(MKL_PARALLEL)
-    if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU" AND NOT APPLE)
-        __mkl_find_library(MKL_THREADING_LIB mkl_gnu_thread)
-    else()
-        __mkl_find_library(MKL_THREADING_LIB mkl_intel_thread)
-    endif()
-else()
-    __mkl_find_library(MKL_THREADING_LIB mkl_sequential)
+
+# Interface
+#
+__mkl_find_library(MKL_INTERFACE_INTEL_32BIT_LIB mkl_intel_lp64)
+__mkl_find_library(MKL_INTERFACE_INTEL_64BIT_LIB mkl_intel_ilp64)
+if(NOT APPLE AND CMAKE_Fortran_COMPILER_LOADED
+             AND CMAKE_Fortran_COMPILER_ID STREQUAL "GNU")
+    __mkl_find_library(MKL_INTERFACE_GF_32BIT_LIB mkl_gf_lp64)
+    __mkl_find_library(MKL_INTERFACE_GF_64BIT_LIB mkl_gf_ilp64)
 endif()
 
-# ScaLAPACK component
-#
-list(FIND MKL_FIND_COMPONENTS SCALAPACK _mkl_find_scalapack)
-if(NOT ${_mkl_find_scalapack} STREQUAL "-1")
-
-    find_package(MPI COMPONENTS CXX)
-    execute_process(COMMAND mpirun --version OUTPUT_VARIABLE MPIRUN_OUTPUT)
-    string(FIND "${MPIRUN_OUTPUT}" "Open MPI" _ompi_pos)
-    
-    # BLACS
-    #
-    if(_ompi_pos STREQUAL "-1")  # MPICH
-        if(APPLE)
-            __mkl_find_library(MKL_BLACS_LIB mkl_blacs_mpich_${_mkl_lp})
-        else()
-            __mkl_find_library(MKL_BLACS_LIB mkl_blacs_intelmpi_${_mkl_lp})
-        endif()
-    else()                      # OpenMPI
-        if(APPLE)
-            message(FATAL_ERROR "Only MPICH is supported on Apple.")
-        endif()
-         __mkl_find_library(MKL_BLACS_LIB mkl_blacs_openmpi_${_mkl_lp})
-    endif()
-    
-    # ScaLAPACK
-    #
-    __mkl_find_library(MKL_SCALAPACK_LIB mkl_scalapack_${_mkl_lp})
-    
-    find_package_handle_standard_args(MKL_SCALAPACK REQUIRED_VARS MKL_BLACS_LIB
-                                                                  MKL_SCALAPACK_LIB
-                                                                  MPI_FOUND)
+# Threading
+#
+__mkl_find_library(MKL_SEQ_LIB mkl_sequential)
+if(NOT APPLE AND (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR
```

Comment:
Whether these flags are required depends on the linker. On Linux, most compilers seem to use the same system provided one, so I think these flags should be added if on Linux in general. The link advisor adds these flags for ICC and PGI as well.
I suggest changing to if(UNIX AND NOT APPLE)